patches/tModLoader/Terraria/ModLoader/IO/WorldIO.cs

```diff
@@ -631,4 +637,123 @@ internal static void EraseWorld(string path, bool cloudSave)
 			SocialAPI.Cloud.Delete(path);
 		}
 	}
+
+	private static TagCompound SaveHeader()
+	{
+		return new TagCompound {
+			["modHeaders"] = SaveModHeaders(),
+			["usedMods"] = SaveUsedMods(),
+			["usedModPack"] = SaveUsedModPack(),
+			["generatedWithMods"] = SaveGeneratedWithMods(),
+		};			
+	}
+
+	private static TagCompound SaveModHeaders()
+	{
+		var modHeaders = new TagCompound();
+
+		var saveData = new TagCompound();
+		foreach (var system in SystemLoader.Systems) {
+			system.SaveWorldHeader(saveData);
+			if (saveData.Count == 0)
+				continue;
+
+			modHeaders[system.FullName] = saveData;
+			saveData = new TagCompound();
+		}
+
+		// preserve data for unloaded systems
+		foreach (var entry in Main.ActiveWorldFileData.ModHeaders) {
+			if (!ModContent.TryFind<ModSystem>(entry.Key, out _))
+				modHeaders[entry.Key] = entry.Value;
+		}
+
+		return modHeaders;
+	}
+
+	internal static void ReadWorldHeader(WorldFileData data)
+	{
+		string path = Path.ChangeExtension(data.Path, ".twld");
+		bool isCloudSave = data.IsCloudSave;
+
+		if (!FileUtilities.Exists(path, isCloudSave))
+			return;
+
+		try {
+			// this code hard-traverses the NBT format to read the just the first nested tag, if preset.
+			// It relies on deterministic tag saving order for the header to be first.
+			// Because the NBT format is not seekable, there is no way to skip reading the entire tag tree in order to find a specific sub-tag at an arbitrary path.
+
+			using Stream stream = isCloudSave ? new MemoryStream(SocialAPI.Cloud.Read(path)) : new FileStream(path, FileMode.Open);
+			using BinaryReader reader = new BigEndianReader(new GZipStream(stream, CompressionMode.Decompress));
+			if (reader.ReadByte() != 10)
+				throw new IOException("Root tag not a TagCompound");
+
+			// ignore root tag name
+			_ = TagIO.ReadTagImpl(8, reader);
+			if (reader.ReadByte() != 10)
+				return; // no header tag
+
+			if ((string)TagIO.ReadTagImpl(8, reader) != "0header")
+				return;
+
+			LoadWorldHeader(data, (TagCompound)TagIO.ReadTagImpl(10, reader));
+		}
+		catch (Exception ex) {
+			Logging.tML.Warn($"Error reading .twld header from: {path} (IsCloudSave={isCloudSave})", ex);
+		}
+	}
+
+	private static void LoadWorldHeader(WorldFileData data, TagCompound tag)
+	{
+		data.ModHeaders = new Dictionary<string, TagCompound>();
+		foreach (var entry in tag.GetCompound("modHeaders")) {
+			string fullname = entry.Key;
+
+			if (ModContent.TryFind<ModSystem>(fullname, out var system)) // handle legacy renames
+				fullname = system.FullName;
+
+			data.ModHeaders[fullname] = (TagCompound)entry.Value;
+		}
```

Comment:
Split this out to `LoadModHeaders` since the other load/save data methods are also split